pkg/apis/core/types.go

```diff
@@ -4142,11 +4142,48 @@ type NodeAddressType string
 
 // These are valid values of node address type
 const (
-	NodeHostName    NodeAddressType = "Hostname"
-	NodeExternalIP  NodeAddressType = "ExternalIP"
-	NodeInternalIP  NodeAddressType = "InternalIP"
-	NodeExternalDNS NodeAddressType = "ExternalDNS"
+	// NodeHostName identifies the node's canonical hostname
+	NodeHostName NodeAddressType = "Hostname"
+
+	// NodeInternalIP identifies an IP address which is assigned to one of the node's
+	// network interfaces. Every node should have at least one address of this type.
+	//
+	// An internal IP is normally expected to be reachable from every other node, but
+	// may or may not be visible to hosts outside the cluster. By default it is
+	// assumed that kube-apiserver can reach node internal IPs, though it is possible
+	// to configure clusters where this is not the case.
+	//
+	// In cloud environments, a NodeInternalIP address generally corresponds to an
```

Comment:
I'd like to avoid terms like "In cloud environments" since it's not clear what a "cloud environment" means and I think this part of the document should be true to all environment anyways.

Reply:
This part (and the corresponding part below for ExternalIP) was kind of written for future CloudProvider implementors: "here's what the other CloudProviders are doing; you should probably do the same." Especially, I was using the words "internal"/"private" vs "external"/"public"/"virtual" to point out specific words that tend to be used in cloud APIs, and show how they usually map to the Kubernetes APIs. This really doesn't apply to the non-CloudProvider use case.

> and I think this part of the document should be true to all environment anyways

I tried to be cloud-vs-non-cloud agnostic in the preceding paragraphs, but in this one I was very explicitly trying to _not_ be agnostic about that. For purposes of node addresses, there's a pretty fundamental distinction between "clusters that use `--cloud-provider`" and "clusters that don't use `--cloud-provider`". Maybe talking about "cloud environments" isn't the right way to describe that (just like saying "bare metal" wasn't).

Reply:
> Maybe talking about "cloud environments" isn't the right way to describe that (just like saying "bare metal" wasn't).

"Cloud provider" is just the most common way of setting these addresses today but this API is independent of it and should not really refer to it IMO.

Reply:
> For purposes of node addresses, there's a pretty fundamental distinction between "clusters that use `--cloud-provider`" and "clusters that don't use `--cloud-provider`"

That can be another angle,from the perspective of "who" sets the IPs, i.e., "... on environments where kubelet delegates the assignation of IP addresses, ....  "

Reply:
> "Cloud provider" is just the most common way of setting these addresses today but this API is independent of it and should not really refer to it IMO.

Hm... ok, I guess I'm mixing up the documentation of `NodeAddressType` with the documentation of `Node.Status.Addresses`. _In theory_, someone could use `NodeAddressType` for something else, but currently, it's only used for `Node.Status.Addresses`, and in that context, it is not independent of cloud providers. (Kubelet requires that either kubelet or an external cloud provider owns `Node.Status.Addresses`.)